src/IECoreDelight/Renderer.cpp

```diff
@@ -389,15 +396,185 @@ IE_CORE_DECLAREPTR( DelightOutput )
 namespace
 {
 
-class DelightShader : public IECore::RefCounted
+// From https://gitlab.com/3Delight/3delight-for-houdini/-/blob/master/osl_utilities.cpp
+enum BasisTypes
+{
+	CONSTANT,
+	LINEAR,
+	MONOTONECUBIC,
+	CATMULLROM
+};
+
+int basisInt( const std::string &basis )
+{
+	if( basis == "constant" )
+	{
+		return BasisTypes::CONSTANT;
+	}
+	if( basis == "linear" )
+	{
+		return BasisTypes::LINEAR;
+	}
+	// `SplinePlug` converts from `monotonecubic` to `bezier`, so we'll never get `monotonecubic`
+
+	return BasisTypes::CATMULLROM;
+}
+
+const OSL::OSLQuery::Parameter *splineValueParameter(
+	const OSL::OSLQuery &query,
+	const std::string &splineParameterName
+)
+{
+	for( size_t i = 0, eI = query.nparams(); i < eI; ++i )
+	{
+		const OSL::OSLQuery::Parameter *p = query.getparam( i );
+
+		if( !boost::starts_with( p->name, splineParameterName ) )
+		{
+			continue;
+		}
+		for( const auto &m : p->metadata )
+		{
+			if(
+				m.name == "widget" &&
+				m.sdefault.size() > 0 &&
+				m.sdefault[0].find( "Ramp" ) != std::string::npos
+			)
+			{
+				return p;
+			}
+		}
+	}
+	return nullptr;
+}
+
+bool find3DelightSplineParameters(
+	const OSL::OSLQuery &query,
+	const std::string &splineParameterName,
+	const OSL::OSLQuery::Parameter * &positionsParameter,
+	const OSL::OSLQuery::Parameter * &valuesParameter,
+	const OSL::OSLQuery::Parameter * &basisParameter
+)
+{
+	positionsParameter = nullptr;
+	valuesParameter = nullptr;
+	basisParameter = nullptr;
+
+	valuesParameter = splineValueParameter( query, splineParameterName );
+
+	if( valuesParameter )
+	{
+		for( size_t i = 0, eI = query.nparams(); i < eI; ++i )
+		{
+			const OSL::OSLQuery::Parameter *p = query.getparam( i );
+			if( p == valuesParameter || !p->type.is_array() || !boost::starts_with( p->name, splineParameterName ) )
+			{
+				continue;
+			}
+			if( p->type.basetype == OSL::TypeDesc::INT && p->type.aggregate == OSL::TypeDesc::SCALAR )
+			{
+				// Here we prefer the `int` value basis parameter because that is the only
+				// basis parameter that is consistently found in all 3delight splines.
+				basisParameter = p;
+			}
+			if( p->type.basetype == OSL::TypeDesc::FLOAT && p->type.aggregate == OSL::TypeDesc::SCALAR )
+			{
+				positionsParameter = p;
+			}
+		}
+	}
+
+	return positionsParameter && valuesParameter && basisParameter;
+}
+
+ConstCompoundDataPtr renameParameters( const std::string &shaderFileName, const ConstCompoundDataPtr &parametersData )
 {
+	CompoundDataPtr newParametersData = new CompoundData();
+	CompoundDataMap &newParameters = newParametersData->writable();
 
+	const CompoundDataMap &oldParameters = parametersData->readable();
+
+	for( const auto &p : oldParameters )
+	{
+		std::string parameter = p.first;
+		DataPtr data = p.second;
+		if(
+			boost::ends_with( parameter, "Positions" ) ||
+			boost::ends_with( parameter, "Values" ) ||
+			boost::ends_with( parameter, "Basis" )
+		)
+		{
+			OSL::OSLQuery query;
```

Comment:
I think OSLQuery is probably slow enough that we don't want to do one every time we hit a spline parameter. See the simple cache we use in OSLShader.cpp.

Reply:
Fixed in https://github.com/GafferHQ/gaffer/pull/5432/commits/0cec0a7dd3f01c812276ce3ac903110f91966b8b